### PR TITLE
Async job

### PR DESCRIPTION
**Pull request by gennad-rocket**

### CODE REVIEW

File: facebookads/objects.py [hunk 1]
```diff
@@ -2512,20 +2514,27 @@ def search(cls, params=None, api=None):
...
+                        search_obj.update(keys[item])
+                        if keys[item]:
+                            ret_val.append(search_obj)
+
+                if 'paging' in response and 'next' in response['paging']:
+                    response = requests.get(response['paging']['next']).json()
+                else:
+                    break
+
         return ret_val
```
Comment (pasha-r) [inline facebookads/objects.py]: It'd be much better not to change anything in this module (think about future merges with upstream).
Just implement specialized methods in asyncobjects.TargetingSearch. As an example look at asyncobjects.TargetingSearch.get_all_countries()

File: facebookads/asyncobjects.py [hunk 2]
```diff
@@ -442,6 +470,9 @@ def get_by_ids(cls, ids, params=None, fields=None, api=None, limit=50):
 
         result = []
         for response in api.get_async_results(cls):
+            if response._request_failed:
+                raise response.last_error()
+
```
Comment (pasha-r) [inline facebookads/asyncobjects.py]: response is an object of AioEdgeIterator type, so response.last_error is an attribute, not a method

File: facebookads/asyncobjects.py [hunk 3]
```diff
@@ -954,3 +1015,191 @@ def __nonzero__(self):
...
             self.remote_read()
         return self[self.Field.async_percent_completion] == 100
+
+    def get_async_status(self):
+        return self[self.Field.async_status]
+
+    def get_async_percent_completion(self):
+        return int(self[self.Field.async_percent_completion])
+
+    def is_failed(self) -> bool:
```
Comment (pasha-r) [inline facebookads/asyncobjects.py]: The Facebook's library is python 2 compatible, so should be our fork of it.
Please replace python3 type hints with docstring type hints (those can be used by Pycharm just as well)

File: facebookads/asyncobjects.py [hunk 4]
```diff
@@ -954,3 +1015,191 @@ def __nonzero__(self):
...
+                        '{} for {}'.format(self.job, self))
+
+            if self.attempt > 8:
+                raise Exception('got no async_status and no '
+                                'report_run_id in {} for {}'.format(self.job, self))
+
+            return True
+        return False
+
+    def check_results(self) -> ([{}], bool):
```
Comment (pasha-r) [inline facebookads/asyncobjects.py]: remove python3 type hint

File: facebookads/asyncobjects.py [hunk 5]
```diff
@@ -954,3 +1015,191 @@ def __nonzero__(self):
...
+        """
+        :type response: facebookadspy.response.AdsAPIResponse
```
Comment (pasha-r) [inline facebookads/asyncobjects.py]: fix docstring type hints

File: facebookads/asyncobjects.py [hunk 6]
```diff
@@ -954,3 +1015,191 @@ def __nonzero__(self):
...
+                else:
+                    time.sleep(1)
+
+                return None, False
+            return self.get_all_results(), True
+
+        else:
+            raise Exception("check_results got unknown stage {} for {}".format(self.stage, self))
+
+    def process_next_stage(self):
```
Comment (pasha-r) [inline facebookads/asyncobjects.py]: Unused method, it would be better just to remove it.

**Comment by gennad-rocket** [comment]
@pasha-r updated

**Pull request merged**